internal/controller/dragonfly_controller.go

```diff
@@ -107,6 +116,81 @@ func (r *DragonflyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		}
 
 		log.Info("Updated resources for object")
+
+		// perform a rollout only if the pod spec has changed
+		var newStatefulset appsv1.StatefulSet
+		if err := r.Get(ctx, client.ObjectKey{Namespace: df.Namespace, Name: df.Name}, &newStatefulset); err != nil {
+			log.Error(err, "could not get statefulset")
+			return ctrl.Result{}, err
+		}
+
+		// check if the pod spec has changed
+		if !reflect.DeepEqual(oldStatefulset.Spec.Template.Spec, newStatefulset.Spec.Template.Spec) {
+			log.Info("Pod spec has changed, performing a rollout")
+			r.EventRecorder.Event(&df, corev1.EventTypeNormal, "Rollout", "Performing a rollout for Dragonfly object")
+
+			// Perform a rollout starting from the replicas
+			// while making sure that they did a full sync before
+			// we go to the next replica
+
+			// Make sure stateful is ready before doing anything
+			var stateful appsv1.StatefulSet
+			if err := r.Get(ctx, client.ObjectKey{Namespace: df.Namespace, Name: df.Name}, &stateful); err != nil {
+				log.Error(err, "could not get statefulset")
+				return ctrl.Result{}, err
+			}
+
+			if stateful.Status.ReadyReplicas != *stateful.Spec.Replicas {
+				log.Info("Statefulset is not ready yet")
+				return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+			}
+
+			// start with replicas
+			var replicas corev1.PodList
+			if err := r.List(ctx, &replicas, client.InNamespace(df.Namespace), client.MatchingLabels(df.Labels), client.MatchingLabels(map[string]string{resources.Role: resources.Replica})); err != nil {
+				log.Error(err, "could not list pods")
+				return ctrl.Result{}, err
+			}
+
+			// delete each replica and wait until the new one is ready
+			// and completed a full sync
+			for _, replica := range replicas.Items {
+				// todo: check what happens when failure happens
+				if err := r.Delete(ctx, &replica); err != nil {
+					log.Error(err, "could not delete replica")
+					return ctrl.Result{}, err
+				}
+
+				// TODO: replace with a proper check
+				time.Sleep(10 * time.Second)
+
+				// wait until the new replica is ready
+				// and completed a full sync
+				var newReplica corev1.Pod
+				if err := r.Get(ctx, client.ObjectKey{Namespace: df.Namespace, Name: replica.Name}, &newReplica); err != nil {
+					log.Error(err, "could not get new replica")
+					return ctrl.Result{}, err
+				}
+
+				// wait until the new replica has done a full sync
+				err := waitForFullSync(ctx, r.Client, &newReplica, 3*time.Minute)
```

Comment:
Do we have to wait here? 
Can we re-queue and check if we are done each time which i think is more inline with programming model and less error prone?
In any case I think that returning an error is problematic as we will not recover. 
So I think we should support  re-queuing the request and proceeding following a re-queue any how.

Reply:
Updated!